### ISSUE TITLE
hash collision in engine leads to incorrect "multiple matches for labels" error

### ISSUE THREAD

**Issue by kevingessner** [prometheus/prometheus#5724]
## Bug Report
**What did you do?**

Run this test case with `promtool test rules`.  This test case is isolated from a series that failed in our production Prometheus instance.

```yaml
evaluation_interval: 1m
tests:
  - interval: 1s
    input_series:
      - series: 'node_namespace_pod:kube_pod_info:{namespace="observability",node="gke-search-infra-custom-96-253440-fli-d135b119-jx00",pod="node-exporter-l454v"}'
        values: '1'
      - series: 'node_cpu_seconds_total{cpu="10",endpoint="https",instance="10.253.57.87:9100",job="node-exporter",mode="idle",namespace="observability",pod="node-exporter-l454v",service="node-exporter"}'
        values: '449'
      - series: 'node_cpu_seconds_total{cpu="35",endpoint="https",instance="10.253.57.87:9100",job="node-exporter",mode="idle",namespace="observability",pod="node-exporter-l454v",service="node-exporter"}'
        values: '449'
      - series: 'node_cpu_seconds_total{cpu="89",endpoint="https",instance="10.253.57.87:9100",job="node-exporter",mode="idle",namespace="observability",pod="node-exporter-l454v",service="node-exporter"}'
        values: '449'
    promql_expr_test:
      - expr: count by(namespace, pod, cpu) (node_cpu_seconds_total{cpu=~".*",job="node-exporter",mode="idle",namespace="observability",pod="node-exporter-l454v"}) * on(namespace, pod) group_left(node) node_namespace_pod:kube_pod_info:{namespace="observability",pod="node-exporter-l454v"}
        eval_time: 4s
        exp_samples:
          - labels: '{cpu="10",namespace="observability",node="gke-search-infra-custom-96-253440-fli-d135b119-jx00",pod="node-exporter-l454v"}'
            value: 1
          - labels: '{cpu="35",namespace="observability",node="gke-search-infra-custom-96-253440-fli-d135b119-jx00",pod="node-exporter-l454v"}'
            value: 1
          - labels: '{cpu="89",namespace="observability",node="gke-search-infra-custom-96-253440-fli-d135b119-jx00",pod="node-exporter-l454v"}'
            value: 1
```

**What did you expect to see?**

A success message, with the three output series correctly calculated.

**What did you see instead? Under which circumstances?**

A test failure:
```
  FAILED:
    expr:'count by(namespace, pod, cpu) (node_cpu_seconds_total{cpu=~".*",job="node-exporter",mode="idle",namespace="observability",pod="node-exporter-l454v"}) * on(namespace, pod) group_left(node) node_namespace_pod:kube_pod_info:{namespace="observability",pod="node-exporter-l454v"}', time:4s, err:multiple matches for labels: grouping labels must ensure unique matches
```

**Investigation**

This failure is due to a hash collision in the VectorBinop code in `engine.go`, discovered by @cristi- .  Specifically, combining the hashes of the LHS and RHS at [L1550](https://github.com/prometheus/prometheus/blob/851131b0740be7291b98f295567a97f32fffc655/promql/engine.go#L1550) results in a collision between two of the series in the test file, `cpu="35"` and `cpu="89"`.  Because the hashes collide there, the result metric is incorrectly reused for two series that are not identical, leading to VectorBinop incorrectly reporting non-unique matches.

You can see the hash-mixing function `(lh ^ rh) + lh` collide in isolation with this code ([playground](https://play.golang.org/p/YKRBZru5XHA)):

```go
package main

import (
	"fmt"
)

func h(l, r uint64) uint64 {
	return (l ^ r) + l
}

func main() {
	fmt.Println(h(4717575973110797488, 6876917458372880354))
	fmt.Println(h(9462489420336518194, 6876917458372880354))
}
```

This would be fixed with a different function for mixing hashes that better preserves entropy, or not combining hashes at all and using the calculated result hash throughout.

**Environment**

* System information: Linux 5.1.0-1.el7.elrepo.x86_64 x86_64
* Prometheus version: reproduced on v2.8.1 and master (851131b0740be7291b98f295567a97f32fffc655)
* Alertmanager version: n/a
* Prometheus configuration file: n/a
* Alertmanager configuration file: n/a
* Logs: n/a

**Comment by brian-brazil** [comment]
No matter what hash function is used, collisions will happen.

**Comment by kevingessner** [comment]
Sure, but I think this xor logic is particularly making them more likely.  `labels.Hash` is a good hash function, and the collision isn't coming from there.

**Comment by juliusv** [comment]
Wow, thanks for tracking this down! Is there a logical / mathematical analysis for why the xor logic makes collisions more likely?

TBH just using hashes without collision detection in the PromQL engine has always seemed dangerous to me, even if the main hash function was improved a while ago :)

**Comment by kevingessner** [comment]
There's some discussion of using xor vs another technique for combining hashes here: https://stackoverflow.com/questions/5889238/why-is-xor-the-default-way-to-combine-hashes

I'm not sure we can combine hashes better than xor, so it might be best to avoid combining them at all.  It looks like that's an optimization to avoid recalculating the 'real' hash for that metric, but using the real hash would remove this collision possibility entirely.

**Comment by juliusv** [comment]
Maybe the weakness is that we do a combo of xor-and-add, where binary addition is basically xor, except for the extra carry bit (so if no two bits are 1 at the same position, it becomes the same as xor). Maybe another binary operator than `+` would be better, hmm...

**Comment by ericajcase** [comment]
Hi!  We are starting to run into this problem on an almost-daily basis in our recording rules, resulting in missed basis.

It comes from automated rule generation Python code that runs on multiple metrics
`{ f"record": f"level1:{metric}:irate", "expr": f"sum without (instance)(irate({metric}[30s]))" + 0*on(app_name, env) group_left(pipeline_id) level1:pipeline:min * 0 }

Is something about this expression making us more vulnerable to collisions?  Is there a better way to add pipline_id?

**Comment by roidelapluie** [comment]
Proposed fix in #6557

**Comment by phakudi** [comment]
@roidelapluie we have a reproducible test case for this too...

```
evaluation_interval: 1m
tests:
  - interval: 15s
    input_series:
      - series: 'A{cityId="66",env="prod",instance="172.31.36.96:8080",job="app-metrics",region="ap-south-1",service="cerebro-listing",serviceability="NON_SERVICEABLE",zoneId="896"}'
        values: '169'
      - series: 'A{cityId="66",env="prod",instance="172.31.39.251:8080",job="app-metrics",region="ap-south-1",service="cerebro-listing",serviceability="SERVICEABLE",zoneId="896"}'
        values: '169'
      - series: 'B{city="Thrissur",cityId="66",env="prod",instance="delivery-metrics-collector-1.swiggy.prod:8080",job="app-metrics",region="ap-south-1",service="metrics-collector",zone="Thr_Mannuthy",zoneId="896"}'
        values: '8'
    promql_expr_test:
      - expr: sum_over_time(A[60s]) + on (cityId, zoneId) group_left(city) (0*(count by (cityId, zoneId, city) (B)))
        eval_time: 60s
        exp_samples:
          - labels: '{cityId="66",env="prod",instance="172.31.36.96:8080",job="app-metrics",region="ap-south-1",service="cerebro-listing",serviceability="NON_SERVICEABLE",zoneId="896",city="Thrissur"}'
            value: 1
          - labels: '{cityId="66",env="prod",instance="172.31.39.251:8080",job="app-metrics",region="ap-south-1",service="cerebro-listing",serviceability="SERVICEABLE",zoneId="896",city="Thrissur"}'
            value: 1
```

**Comment by astral303** [comment]
Hit this in production, making some charts no longer return any data. This explains why excluding a certain instance inexplicably makes queries succeed again -- it's removing the collision.

**Comment by roidelapluie** [comment]
Fixed in #7058
